### PR TITLE
[IDE] Make the folding margin keyboard accessible

### PR DESCRIPTION
**Pull request by iainx**
When in the text editor, use Alt-Tab to focus the margins.
Tab, Shift+Tab, Left or Right to select the focus margin (Only folding margin
accepts focus at the moment).
Up and Down selects the fold
Space folds or unfolds the selected fold

Fixes VSTS #515126

### CODE REVIEW

**Comment by iainx** [comment]
fold keyboard

File: main/src/addins/MonoDevelop.SourceEditor2/Mono.TextEditor/Gui/FoldMarkerMargin.cs [hunk 1]
```diff
@@ -487,7 +494,97 @@ internal protected override void Draw (Cairo.Context cr, Cairo.Rectangle area, D
 				}
 			}
 		}
-	}
+
+		FoldSegment[] focusSegments;
+		int focusedIndex;
+		protected internal override bool SupportsItemCommands => true;
+        protected internal override bool HandleItemCommand(ItemCommand command)
```
Comment (slluis) [inline main/src/addins/MonoDevelop.SourceEditor2/Mono.TextEditor/Gui/FoldMarkerMargin.cs]: Spaces!

**Comment by slluis** [comment]
@mkrueger can you please review?

File: main/src/addins/MonoDevelop.SourceEditor2/Mono.TextEditor/Gui/TextArea.cs [hunk 2]
```diff
@@ -668,37 +668,185 @@ void IMCommit (object sender, Gtk.CommitArgs ca)
...
+		void FocusIn ()
 		{
-			var result = base.OnFocusInEvent (evnt);
 			imContextNeedsReset = true;
 			IMContext.FocusIn ();
 			RequestResetCaretBlink ();
 			Document.CommitLineUpdate (Caret.Line);
+		}
+
+        protected override bool OnFocusInEvent (EventFocus evnt)
```
Comment (mkrueger) [inline main/src/addins/MonoDevelop.SourceEditor2/Mono.TextEditor/Gui/TextArea.cs]: Again - spaces. That terrible bug is fixed in theory - but we need the editorconfig support merged in master for that.

15.6 however works.

File: main/src/addins/MonoDevelop.SourceEditor2/Mono.TextEditor/Gui/TextViewMargin.cs [hunk 3]
```diff
@@ -2451,7 +2451,7 @@ protected internal override void MouseReleased (MarginMouseEventArgs args)
 			base.MouseReleased (args);
 		}
 
-		CodeSegmentPreviewWindow previewWindow = null;
+        CodeSegmentPreviewWindow previewWindow = null;
```
Comment (mkrueger) [inline main/src/addins/MonoDevelop.SourceEditor2/Mono.TextEditor/Gui/TextViewMargin.cs]: Spaces again.

**Comment by mkrueger** [comment]
Overall funny feature I didn't implement it that way - but I think it's a nice addition.

File: main/src/addins/MonoDevelop.SourceEditor2/Mono.TextEditor/Gui/TextArea.cs (hunk 2 above)
Comment (iainx) [reply main/src/addins/MonoDevelop.SourceEditor2/Mono.TextEditor/Gui/TextArea.cs]: argh...

**Comment by iainx** [comment]
Fixed spaces and added keyboard focus for the icon margin as well
margins-focus

**Comment by iainx** [comment]
@slluis done

**Comment by iainx** [comment]
This has been okayed by QA

**Pull request merged**